ndhistogram/src/axis/cyclic.rs

```diff
@@ -0,0 +1,199 @@
+/// ```
+/// use ndhistogram::{ndhistogram, Histogram};
+/// use ndhistogram::axis::{Axis, BinInterval, Cyclic};
+/// let mut hist = ndhistogram!(Cyclic::new(4, 0.0, 360.0));
+/// hist.fill(& 45.0         ); // Add entry at 45 degrees
+/// hist.fill(&(45.0 + 360.0)); // Add entry at 45 degrees + one whole turn
+/// hist.fill(&(45.0 - 360.0)); // Add entry at 45 degrees + one whole turn backwards
+/// // All 3 above entries end up in the same bin
+/// assert_eq!(hist.value(&45.0), Some(&3.0));
+/// // Lookup also wraps around
+/// assert_eq!(hist.value(&(45.0 + 360.0)), Some(&3.0));
+/// assert_eq!(hist.value(&(45.0 - 360.0)), Some(&3.0));
+/// ```
+/// Time of day
+/// ```
+/// use ndhistogram::{ndhistogram, Histogram};
+/// use ndhistogram::axis::{Axis, BinInterval, Cyclic};
+/// let bins_per_day = 24;
+/// let hours_per_bin = 1;
+/// let start_at_zero = 0;
+/// let four_pm = 16;
+/// let mut hist = ndhistogram!(Cyclic::with_step_size(
+///     bins_per_day, start_at_zero, hours_per_bin
+/// ));
+/// hist.fill(&40);                               // The 40th hour of the week ...
+/// assert_eq!(hist.value(&four_pm), Some(&1.0)); // ... is at 4 pm.
+/// ````
+#[derive(Default, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug, Serialize, Deserialize)]
+pub struct Cyclic<T = f64> {
+    axis: Uniform<T>,
+}
+
+impl<T> Cyclic<T>
+where
+    T: PartialOrd + Num + NumCast + NumOps + Copy,
+{
+    /// Create a wrap-around axis with `nbins` uniformly-spaced bins in the range `[low, high)`.
+    ///
+    /// Only implemented for [Float]. Use [Cyclic::with_step_size] for integers.
+    ///
+    /// # Panics
+    /// Panics under the same conditions as [Uniform::new].
+    pub fn new(nbins: usize, low: T, high: T) -> Self
+    where
+        T: Float,
+    {
+        Self {
+            axis: Uniform::new(nbins, low, high),
+        }
+    }
+
+    /// Create a wrap-around axis with `nbins` uniformly-spaced bins in the range `[low, low+num*step)`.
+    /// # Panics
+    /// Panics under the same conditions as [Uniform::new].
+    pub fn with_step_size(nbins: usize, low: T, step: T) -> Self {
+        Self {
+            axis: Uniform::with_step_size(nbins, low, step),
+        }
+    }
+}
+
+impl<T> Cyclic<T> {
+    /// Low edge of axis (excluding wrap-around) // TODO or should this be - infinity?
+    pub fn low(&self) -> &T {
+        self.axis.low()
+    }
+    /// High edge of axis (excluding wrap-around) // TODO or should this be + infinity?
+    pub fn high(&self) -> &T {
+        self.axis.high()
+    }
+}
+
+// TODO integers?
+impl<T: PartialOrd + NumCast + NumOps + Copy> Axis for Cyclic<T> {
+    type Coordinate = T;
+    type BinInterval = BinInterval<T>;
+
+    // TODO optimize by using division instead of looping?
+    fn index(&self, coordinate: &Self::Coordinate) -> Option<usize> {
+        let (mut x, hi, lo) = (*coordinate, *self.axis.high(), *self.axis.low());
+        let range = hi - lo;
+        while x >= hi {
```

Comment:
I think that this while will loop infinitely if coordinate is `inf`. As you point out in the TODO comment. It's better to do something else.

I think that the modulus operator will work for both integers and floats. Something like:
```
        x = (x - lo) % range;
        if x < <T as Zero>::zero() {
            x = range + x;
        }
        x = x + lo;
```